src/query/graphite/graphite/timespec.go

```diff
@@ -108,9 +137,148 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
 		return time.Unix(n, 0).UTC(), nil
 	}
 
+	ref, offset := s, ""
+	if strings.Contains(s, "+") {
+		stringSlice := strings.Split(s, "+")
+		ref, offset = stringSlice[0], stringSlice[1]
+		offset = "+" + offset
+	} else if strings.Contains(s, "-") {
+		stringSlice := strings.Split(s, "-")
+		ref, offset = stringSlice[0], stringSlice[1]
+		offset = "-" + offset
+	}
+	parsedReference, err := ParseTimeReference(ref, now)
+	if err == nil {
+		parsedOffset, err := ParseOffset(offset)
+		if err == nil {
+			return parsedReference.Add(parsedOffset), nil
+		}
+	}
+
 	return now, err
 }
 
+func ParseTimeReference(ref string, now time.Time) (time.Time, error) {
+	var (
+		hour    = now.Hour()
+		minute  = now.Minute()
+		refDate time.Time
+	)
+
+	if ref == "" || ref == "now" {
+		return now, nil
+	}
+
+	// check if the time ref fits an absolute time format
+	for _, format := range formats {
+		t, err := time.Parse(format, ref)
+		if err == nil {
+			return time.Date(t.Year(), t.Month(), t.Day(), hour, minute, 0, 0, now.Location()), nil
+		}
+	}
+
+	rawRef := ref
+
+	// Time of Day Reference
+	i := strings.Index(ref, ":")
+	if 0 < i && i < 3 {
+		newHour, err := strconv.ParseInt(ref[:i], 10, 0)
+		if err != nil {
+			return time.Time{}, err
+		}
+		hour = int(newHour)
+		if len(ref) >= i+3 {
+			newMinute, err := strconv.ParseInt(ref[i+1:i+3], 10, 32)
+			if err != nil {
+				return time.Time{}, err
+			}
+			minute = int(newMinute)
+			ref = ref[i+3:]
+		}
+
+		if len(ref) >= 2 {
+			if ref[:2] == "am" {
+				ref = ref[2:]
+			} else if ref[:2] == "pm" {
+				hour = (hour + 12) % 24
+				ref = ref[2:]
+			}
+		}
+	}
+
+	// Xam or XXam
+	i = strings.Index(ref, "am")
+	if 0 < i && i < 3 {
+		newHour, err := strconv.ParseInt(ref[:i], 10, 32)
+		if err != nil {
+			return time.Time{}, err
+		}
+		hour = int(newHour)
+		minute = 0
+		ref = ref[i+2:]
+	}
+
+	// Xpm or XXpm
+	i = strings.Index(ref, "pm")
+	if 0 < i && i < 3 {
+		newHour, err := strconv.ParseInt(ref[:i], 10, 32)
+		if err != nil {
+			return time.Time{}, err
+		}
+		hour = int((newHour + 12) % 24)
+		minute = 0
+		ref = ref[i+2:]
+	}
+
+	if strings.HasPrefix(ref, "noon") {
+		hour, minute = 12, 0
+		ref = ref[4:]
+	} else if strings.HasPrefix(ref, "midnight") {
+		hour, minute = 0, 0
+		ref = ref[8:]
+	} else if strings.HasPrefix(ref, "teatime") {
+		hour, minute = 16, 0
+		ref = ref[7:]
+	}
+
+	refDate = time.Date(now.Year(), now.Month(), now.Day(), hour, minute, 0, 0, now.Location())
+
+	// Day reference
+	if ref == "yesterday" {
```

Comment:
Should these match against the initial `ref` before it's been shortened?

Reply:
Yes. Fixed.

Reply:
No, because this function needs to be able to parse "noon yesterday" which will actually be "noonyesterday" when passed in because whitespace gets stripped. Parsing the unshortened ref could break this.